torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py

```diff
@@ -0,0 +1,130 @@
+"""JIT C++ strings into executables.
+
+1) Enable timeit and Callgrind collection for C++ snippets.
+2) Allow Timer.collect_callgrind to be used on earlier versions of PyTorch
+"""
+import atexit
+import os
+import re
+import shutil
+import textwrap
+import threading
+from types import ModuleType
+from typing import cast, List, Optional, Union, TYPE_CHECKING
+import uuid
+
+if TYPE_CHECKING:
+    from typing import Protocol
+else:
+    Protocol = object
+
+import torch
+from torch.utils import cpp_extension
+
+
+SOURCE_ROOT = os.path.split(os.path.abspath(__file__))[0]
+with open(os.path.join(SOURCE_ROOT, "timer_timeit_template.cpp"), "rt") as f:
+    CXX_TIMEIT_TEMPLATE = f.read()
+
+
+with open(os.path.join(SOURCE_ROOT, "timer_callgrind_template.cpp"), "rt") as f:
+    CXX_CALLGRIND_TEMPLATE = f.read()
+
+
+BUILD_ROOT = os.path.join(
+    torch._appdirs.user_cache_dir(appname="benchmark_utils_jit"),
+    f"build_{uuid.uuid4()}".replace("-", "")
+)
+os.makedirs(BUILD_ROOT, exist_ok=True)
+WRITE_SOURCE_LOCK = threading.Lock()
+atexit.register(shutil.rmtree, BUILD_ROOT)
+
+
+CXX_FLAGS = torch.__config__._cxx_flags().strip().split()
+if "-g" not in CXX_FLAGS:
+    CXX_FLAGS.append("-g")
+
+
+# load will automatically search /usr/include, but not conda include.
+extra_include_paths: List[str] = []
+conda_prefix = os.getenv("CONDA_PREFIX")
+if conda_prefix is not None:
+    extra_include_paths = [os.path.join(conda_prefix, "include")]
+extra_include_paths.append(SOURCE_ROOT)
+
+
+# MyPy has no way of knowing the methods on a C module, so we must
+# define API classes which mirror the PyBind11 methods.
+class TimeitModuleType(Protocol):
+    def timeit(self, number: int) -> float:
+        ...
```

Comment:
The more conventional way to do this is to define a `pyi` file for the C bindings. You can see examples of this in `torch/_C`. If you're doing the pybind11 by literally grabbing this module and then stuffing extra bindings on it... well, don't do that, it's super confusing for code readers who are like, "wait, where the heck did this binding come from"